### ISSUE TITLE
nuget odata support

### ISSUE THREAD

**Issue by vradzius** [Hoffs/vscode-versionlens#11]
Support odata nuget package repository format.
npm odata-client can be used for that
This would add support for private nuget repositories which uses nuget.server from https://github.com/NuGet/NuGet.Server.
This can be tested on https://www.nuget.org/api/v2/ endpoint

**Comment by vradzius** [comment]
Suggested changes for this to work:
file: src\providers\dotnet\nugetResolvers\odataClientResolver.js
```js
import appContrib from 'common/appContrib';

// This is a required NugetV3 api
// Uses PackageBaseAddress API
export async function odataClientResolver(serviceUrl, packageName) {
  const odata = require('odata-client');
  // From SearchAutocompleteService
  if (!serviceUrl.endsWith('/')) {
    serviceUrl = `${serviceUrl}/`;
  }
  const query = odata({service: serviceUrl, resources: 'Packages'});
  const filterExp = `Id eq '${packageName}'`;
  const response = await query.filter(filterExp).select('Version').get();
  if (response.statusCode != 200) {
    throw {
      status: response.statusCode,
      responseText: response.body
    };
  }
  const convert = require('xml-js');
  const data = convert.xml2js(response.body);
  let versions = [];
  if (!data) {
    throw { status: 404 };
  } else {
      const packages = data.elements[0].elements.filter(e=>e.name==='entry').map(e=>e.elements.filter(e=>e.name==='m:properties')[0]);
      if(packages.length === 0)
      {
        throw { status: 404 };
      }
      else {
        return packages.map(e=>e.elements.filter(e=>e.name==='d:Version')[0]).map(e=>e.elements[0].text).reverse();
      }
  }
  return versions.reverse();
}
```
changed partial file: src\providers\dotnet\nugetClient.js 
```js
import { searchQueryServiceResolver } from './nugetResolvers/searchQueryServiceResolver';
import { odataClientResolver } from './nugetResolvers/odataClientResolver';

// From https://docs.microsoft.com/en-us/nuget/api/overview
// Sorted in order that we want to get them.
const nugetServiceResolvers = [
  { type: 'OData/2.0.0', resolver: odataClientResolver },
  { type: 'PackageBaseAddress/3.0.0', resolver: packageBaseAddressResolver },
  { type: 'SearchAutocompleteService', resolver: searchAutocompleteServiceResolver },
  { type: 'SearchAutocompleteService/3.0.0-beta', resolver: searchAutocompleteServiceResolver },
  { type: 'SearchAutocompleteService/3.0.0-rc', resolver: searchAutocompleteServiceResolver },
  { type: 'RegistrationsBaseUrl', resolver: registrationsBaseUrlResolver },
  { type: 'RegistrationsBaseUrl/3.0.0-beta', resolver: registrationsBaseUrlResolver },
  { type: 'RegistrationsBaseUrl/3.0.0-rc', resolver: registrationsBaseUrlResolver },
  { type: 'RegistrationsBaseUrl/3.4.0', resolver: registrationsBaseUrlResolver },
  { type: 'RegistrationsBaseUrl/3.6.0', resolver: registrationsBaseUrlResolver },
  { type: 'SearchQueryService', resolver: searchQueryServiceResolver },
  { type: 'SearchQueryService/3.0.0-beta', resolver: searchQueryServiceResolver },
  { type: 'SearchQueryService/3.0.0-rc', resolver: searchQueryServiceResolver },
];

async function getVersionResolverFromIndex(index) {
  const httpRequest = require('request-light');
  const indexResponse = await httpRequest.xhr({ url: index });

  if (indexResponse.status != 200) {
    throw {
      status: indexResponse.status,
      responseText: indexResponse.responseText
    };
  }
  if (indexResponse.responseText.startsWith('<?xml version="1.0" encoding="utf-8"?>'))
  {
    const serviceResolver = nugetServiceResolvers[0];
    return { index, url: index, ...serviceResolver};
  }
  else
  {
    const indexData = JSON.parse(indexResponse.responseText);

    // Go through the list with each resolver, as we prefer to get a resolver
    // that is towards the start of the service resolver list.
    // More efficient would be to just pick first acceptable resolver, but thats not ideal.
    for (const serviceResolver of nugetServiceResolvers) {
      for (const resource of indexData.resources) {
        if (resource['@type'] === serviceResolver.type) {
          return { index, url: resource['@id'], ...serviceResolver }; // return index + url + resolver
        }
      }
    }
  }

  throw { status: 404, responseText: 'No services with available resolvers found in nuget indexes.' };
}
```

**Comment by Hoffs** [comment]
I'll look into this, do you not want to submit PR?

**Comment by Hoffs** [comment]
Published new version 0.25.4 which should include support for this.

**Comment by vradzius** [comment]
Dėkoju ;)
